File: frontend/src/api/teams.ts
import { api } from "./client";

export type TeamVisibility = "public" | "private";

export interface TeamListItem {
  id: string;
  name: string;
  visibility: TeamVisibility;
  member_count: number;
  max_members: number;
  has_space: boolean;
  requires_password: boolean;
}

export interface TeamMember {
  user_id: string;
  display_name: string;
  is_leader: boolean;
  has_avatar: boolean;
  joined_at: string;
}

export interface TeamDetail extends TeamListItem {
  leader_user_id: string;
  members: TeamMember[];
  created_at: string;
}

export interface JoinRequest {
  id: string;
  team_id: string;
  user_id: string;
  display_name: string;
  status: "pending" | "accepted" | "rejected" | "cancelled";
  message: string | null;
  created_at: string;
}

export const listTeams = () => api.get<TeamListItem[]>("/teams");
export const getTeam = (id: string) => api.get<TeamDetail>(`/teams/${id}`);

export const createTeam = (input: {
  name: string;
  visibility: TeamVisibility;
  join_password?: string | null;
}) => api.post<TeamDetail>("/teams", input);

export const updateTeam = (
  id: string,
  input: {
    name?: string;
    visibility?: TeamVisibility;
    join_password?: string | null;
    clear_password?: boolean;
  },
) => api.patch<TeamDetail>(`/teams/${id}`, input);

export const joinTeam = (id: string, password?: string) =>
  api.post<TeamDetail>(`/teams/${id}/join`, { password: password ?? null });

export const leaveTeam = (teamId: string, userId: string) =>
  api.delete<{ message: string }>(`/teams/${teamId}/members/${userId}`);

export const transferLeadership = (teamId: string, userId: string) =>
  api.post<TeamDetail>(`/teams/${teamId}/leader`, { user_id: userId });

export const requestToJoin = (teamId: string, message?: string) =>
  api.post<JoinRequest>(`/teams/${teamId}/join-requests`, { message: message ?? null });

export const listJoinRequests = (teamId: string) =>
  api.get<JoinRequest[]>(`/teams/${teamId}/join-requests`);

export const acceptJoinRequest = (teamId: string, requestId: string) =>
  api.post<TeamDetail>(`/teams/${teamId}/join-requests/${requestId}/accept`);

export const rejectJoinRequest = (teamId: string, requestId: string) =>
  api.post<{ message: string }>(`/teams/${teamId}/join-requests/${requestId}/reject`);

File: backend/app/schemas/teams.py
"""Request and response models for parties."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.team import JoinRequestStatus, TeamVisibility


class TeamMemberResponse(BaseModel):
    user_id: UUID
    display_name: str
    is_leader: bool
    has_avatar: bool
    joined_at: datetime


class TeamListItem(BaseModel):
    """The party browser. Deliberately shows no member identities."""

    id: UUID
    name: str
    visibility: TeamVisibility
    member_count: int
    max_members: int
    has_space: bool
    #: Whether joining needs a password, so the UI can prompt for one. Says
    #: nothing about what the password is.
    requires_password: bool


class TeamDetailResponse(BaseModel):
    id: UUID
    name: str
    visibility: TeamVisibility
    member_count: int
    max_members: int
    has_space: bool
    requires_password: bool
    leader_user_id: UUID
    members: list[TeamMemberResponse]
    created_at: datetime


class CreateTeamRequest(BaseModel):
    name: str = Field(min_length=3, max_length=32)
    visibility: TeamVisibility = TeamVisibility.PUBLIC
    #: Private parties only. Omitted means request-to-join is the only route in.
    join_password: str | None = Field(default=None, min_length=6, max_length=128)


class UpdateTeamRequest(BaseModel):
    name: str | None = Field(default=None, min_length=3, max_length=32)
    visibility: TeamVisibility | None = None
    join_password: str | None = Field(default=None, min_length=6, max_length=128)
    clear_password: bool = False


class JoinTeamRequest(BaseModel):
    password: str | None = Field(default=None, max_length=128)


class TransferLeadershipRequest(BaseModel):
    user_id: UUID


class CreateJoinRequest(BaseModel):
    message: str | None = Field(default=None, max_length=280)


class JoinRequestResponse(BaseModel):
    id: UUID
    team_id: UUID
    user_id: UUID
    display_name: str
    status: JoinRequestStatus
    message: str | None
    created_at: datetime
